pkg/obs/nodeserver.go

```diff
@@ -17,38 +17,147 @@ limitations under the License.
 package obs
 
 import (
+	"fmt"
 	"github.com/container-storage-interface/spec/lib/go/csi"
+	"github.com/huaweicloud/huaweicloud-csi-driver/pkg/obs/services"
 	"github.com/huaweicloud/huaweicloud-csi-driver/pkg/utils/metadatas"
 	"github.com/huaweicloud/huaweicloud-csi-driver/pkg/utils/mounts"
+	"github.com/kubernetes-csi/csi-lib-utils/protosanitizer"
 	"golang.org/x/net/context"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
+	log "k8s.io/klog/v2"
+	"net/http"
+	"os"
+	"path"
+	"path/filepath"
 )
 
 type nodeServer struct {
-	Driver   *Driver
-	Mount    mounts.IMount
-	Metadata metadatas.IMetadata
+	Driver      *Driver
+	Mount       mounts.IMount
+	Metadata    metadatas.IMetadata
+	MountClient http.Client
 }
 
+const (
+	credentialFile = "/host/etc/passwd-obsfs"
+	credential     = "/etc/passwd-obsfs"
+	defaultOpts    = "-o big_writes -o max_write=131072 -o use_ino"
+	perm           = 0600
+)
+
 func (ns *nodeServer) NodeStageVolume(_ context.Context, req *csi.NodeStageVolumeRequest) (
 	*csi.NodeStageVolumeResponse, error) {
+	log.Infof("NodeStageVolume: called with args %v", protosanitizer.StripSecrets(*req))
 	return nil, status.Error(codes.Unimplemented, "")
 }
 
 func (ns *nodeServer) NodeUnstageVolume(_ context.Context, req *csi.NodeUnstageVolumeRequest) (
 	*csi.NodeUnstageVolumeResponse, error) {
+	log.Infof("NodeStageVolume: called with args %v", protosanitizer.StripSecrets(*req))
 	return nil, status.Error(codes.Unimplemented, "")
 }
 
 func (ns *nodeServer) NodePublishVolume(_ context.Context, req *csi.NodePublishVolumeRequest) (
 	*csi.NodePublishVolumeResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+	log.Infof("NodePublishVolume: called with args %v", protosanitizer.StripSecrets(*req))
+	capability := req.GetVolumeCapability()
+	volumeID := req.GetVolumeId()
+	targetPath := req.GetTargetPath()
+	if err := nodePublishValidation(capability, volumeID, targetPath); err != nil {
+		return nil, err
+	}
+
+	credentials := ns.Driver.cloud
+	volume, err := services.GetParallelFSBucket(credentials, volumeID)
+	if err != nil {
+		return nil, err
+	}
+	log.Infof("NodePublishVolume: volume detail: %s", protosanitizer.StripSecrets(volume))
+
+	notMnt, err := ns.Mount.IsLikelyNotMountPointAttach(targetPath)
+	if err != nil {
+		return nil, err
+	}
+	if !notMnt {
+		log.Infof("NodePublishVolume: %s has already mounted", targetPath)
+		return &csi.NodePublishVolumeResponse{}, nil
+	}
+	if err := ns.Mount.MakeDir(targetPath); err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to make dir: %s, error: %v", targetPath, err)
+	}
+	if err := createCredentialFile(ns.Driver.cloud.Global.AccessKey, ns.Driver.cloud.Global.SecretKey); err != nil {
+		return nil, err
+	}
+
+	mntCmd := fmt.Sprintf("obsfs %s %s -o url=obs.%s.%s -o passwd_file=%s %s", volume.BucketName, targetPath,
```

Comment:
Passing commands is not a good way, please refactor later.

Reply:
it will be modified in another PR.